consensus/cryptarchia-sync/network/src/outgoing_stream.rs

```diff
@@ -0,0 +1,187 @@
+use std::future::Future;
+
+use cryptarchia_engine::Slot;
+use futures::{future::BoxFuture, stream::FuturesUnordered, AsyncWriteExt, FutureExt, StreamExt};
+use libp2p::PeerId;
+use libp2p_stream::Control;
+use tokio::{
+    sync::mpsc::Sender,
+    time::{error::Elapsed, timeout, Duration},
+};
+use tracing::{error, info};
+
+use crate::{
+    behaviour::{SyncError, SYNC_PROTOCOL},
+    membership::ConnectedPeers,
+    messages::{SyncDirection, SyncRequest, SyncResponse},
+    sync_utils,
+};
+
+type TipResult = Result<Result<(PeerId, Slot), SyncError>, Elapsed>;
+
+/// Find the peer with the highest tip among the given peers.
+pub async fn select_best_peer_for_sync(
+    control: Control,
+    peers: Vec<PeerId>,
+    local_peer_id: PeerId,
+) -> Result<PeerId, SyncError> {
+    let filtered_peers = filter_valid_peers(peers, local_peer_id)?;
+    let tip_requests = create_tip_requests(filtered_peers, &control);
+    let best_peer = find_highest_tip_peer(tip_requests).await?;
+    Ok(best_peer)
+}
+
+fn filter_valid_peers(peers: Vec<PeerId>, local_peer_id: PeerId) -> Result<Vec<PeerId>, SyncError> {
+    let filtered: Vec<PeerId> = peers
+        .into_iter()
+        .filter(|&id| id != local_peer_id)
+        .collect();
+
+    if filtered.is_empty() {
+        return Err(SyncError::NoPeersAvailable);
+    }
+
+    info!(num_peers = filtered.len(), "Selecting best peer for sync");
+    Ok(filtered)
+}
+
+fn create_tip_requests(
+    peers: Vec<PeerId>,
+    control: &Control,
+) -> FuturesUnordered<impl Future<Output = TipResult>> {
+    peers
+        .into_iter()
+        .map(|peer_id| {
+            let control = control.clone();
+            async move {
+                timeout(
+                    Duration::from_secs(5),
+                    request_tip_from_peer(peer_id, control),
+                )
+                .await
+            }
+        })
+        .collect::<FuturesUnordered<_>>()
+}
+
+async fn find_highest_tip_peer(
+    mut tip_requests: FuturesUnordered<
+        impl Future<Output = Result<Result<(PeerId, Slot), SyncError>, Elapsed>>,
+    >,
+) -> Result<PeerId, SyncError> {
+    let mut best_peer = None;
+    let mut highest_tip = Slot::from(0);
+
+    while let Some(result) = tip_requests.next().await {
+        match result {
+            Ok(Ok((peer_id, tip))) => {
+                if tip > highest_tip {
+                    highest_tip = tip;
+                    best_peer = Some(peer_id);
+                }
+            }
+            Ok(Err(e)) => {
+                error!(error = %e, "Error requesting tip from peer");
+            }
+            Err(_) => {
+                error!("Timeout while requesting tip from peer");
+            }
+        }
+    }
+
+    best_peer.ok_or(SyncError::NoPeersAvailable)
+}
+pub async fn stream_blocks_from_peer(
+    peer_id: PeerId,
+    control: &mut Control,
+    direction: SyncDirection,
+    response_sender: Sender<(Vec<u8>, PeerId)>,
```

Comment:
I think we should use at least aliases, what is the `(Vec<u8>, PeerId)` refer to? `BlockBytes` & `Peer` I guess?